pkg/flow/logging/logger.go

```diff
@@ -3,77 +3,84 @@ package logging
 import (
 	"fmt"
 	"io"
+	"log/slog"
 	"sync"
 	"time"
 
-	"github.com/go-kit/log"
-	"github.com/go-kit/log/level"
 	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/internal/slogadapter"
 	"github.com/grafana/loki/pkg/logproto"
 	"github.com/prometheus/common/model"
 )
 
-// Logger implements the github.com/go-kit/log.Logger interface. It supports
-// being dynamically updated at runtime.
+// Logger is the logging subsystem of Flow. It supports being dynamically
+// updated at runtime.
 type Logger struct {
-	w io.Writer
+	inner io.Writer // Writer passed to New.
 
-	mut sync.RWMutex
-	l   log.Logger
+	level   *slog.LevelVar // Current configured level.
+	format  *formatVar     // Current configured format.
+	writer  *writerVar     // Current configured multiwriter (inner + write_to).
+	handler *handler       // Handler which handles logs.
 }
 
 // New creates a New logger with the default log level and format.
 func New(w io.Writer, o Options) (*Logger, error) {
-	inner, err := buildLogger(w, o)
-	if err != nil {
-		return nil, err
+	var (
+		leveler slog.LevelVar
+		format  formatVar
+		writer  writerVar
+	)
+
+	l := &Logger{
+		inner: w,
+
+		level:  &leveler,
+		format: &format,
+		writer: &writer,
```

Comment:
We only ever use these as a way to set the pointer's value, which is mirrored to the handler.

Should we remove these top-level fields and use the setters directly on the handler's fields?

Reply:
I'd prefer not to; since WithAttrs and WithGroup are expected to propagate down the pointers down to their caller, it feels cleaner to me if they originate from something which isn't a handler and that the handlers are purely pulling the values rather than being set.

Reply:
Ok, sounds fair!